gossip/src/duplicate_shred_handler.rs

```diff
@@ -139,21 +170,59 @@ impl DuplicateShredHandler {
                 return false;
             }
         }
-        // Also skip frozen slots or slots with an older proof than me.
+        // Also skip frozen slots or slots with a different proof than me.
         match self.chunk_map.get(&slot) {
-            Some(SlotStatus::Frozen) => {
-                return false;
-            }
-            Some(SlotStatus::UnfinishedProof(slot_map)) => {
-                if let Some(proof_chunkmap) = slot_map.peek(&data.from) {
-                    if proof_chunkmap.wallclock < data.wallclock {
-                        return false;
+            Some(SlotStatus::Frozen) => false,
+            Some(SlotStatus::UnfinishedProof(slot_map)) => match slot_map.get(&data.from) {
+                None => self.check_has_enough_stake_and_cleanup(&slot, &data.from),
+                Some(proof_chunkmap) => proof_chunkmap.wallclock == data.wallclock,
+            },
+            None => true,
+        }
+    }
+
+    fn check_has_enough_stake_and_cleanup(&mut self, slot: &Slot, newkey: &Pubkey) -> bool {
+        match self.chunk_map.get_mut(slot) {
+            None => true,
+            Some(SlotStatus::Frozen) => false,
+            Some(SlotStatus::UnfinishedProof(chunk_map)) => {
+                if chunk_map.len() < MAX_PUBKEY_PER_SLOT {
+                    return true;
+                }
+                match &mut self.cached_staked_nodes {
+                    Some(cached_staked_nodes) => {
+                        let newkey_stake =
+                            cached_staked_nodes.get(newkey).copied().unwrap_or_default();
+                        // Ignore new pubkey without any stake.
+                        if newkey_stake == 0u64 {
+                            return false;
+                        }
+                        let mut heap = BinaryHeap::new();
+                        // Push new pubkey in first so it's compared with those in the table.
+                        heap.push(Reverse::<(u64, Pubkey)>((newkey_stake, *newkey)));
+                        for oldkey in chunk_map.keys() {
+                            let oldkey_stake =
+                                cached_staked_nodes.get(oldkey).copied().unwrap_or_default();
+                            heap.push(Reverse::<(u64, Pubkey)>((oldkey_stake, *oldkey)));
+                        }
+                        let mut newkey_popped = false;
+                        for _ in 0..(MAX_PUBKEY_PER_SLOT / 10) {
+                            if let Some(Reverse::<(u64, Pubkey)>((_, key))) = heap.pop() {
+                                if &key == newkey {
+                                    newkey_popped = true;
+                                } else {
+                                    chunk_map.remove(&key);
+                                }
+                            }
+                        }
```

Comment:
This is unnecessarily completed.
I think it would be simpler if:
* always insert the entry regardless of stake.
* once the insert is done, if the number of entries >= 2 * times intended capacity, evict half of entries with lowest stake from that hash-map.

Reply:
That's a good point, changed.